### PR TITLE
added precondition of matmul

### PR DESCRIPTION
**Pull request by SumanSudhir**
Listed in #517

- [ ]  precondition for matmul

### CODE REVIEW

File: Sources/TensorFlow/Operators/Math.swift [hunk 1]
```diff
@@ -2748,6 +2748,9 @@ public func matmul<Scalar: Numeric>(
     _ rhs: Tensor<Scalar>,
     transposed transposeRhs: Bool = false
 ) -> Tensor<Scalar> {
+    precondition(lhs.rank >= 2, "The input must have rank greater than or equal to 2.")
+    precondition(rhs.rank >= 2, "The input must have rank greater than or equal to 2.")
+    precondition(lhs.shape[1] == rhs.shape[0] , "Dimesion must be equal,but are \(lhs.shape[1]) and \(rhs.shape[0]).")
```
Comment (Shashi456) [inline Sources/TensorFlow/Operators/Math.swift]: The spelling of dimension and it would be plural in this case.

Comment (SumanSudhir) [reply Sources/TensorFlow/Operators/Math.swift]: resolved

**Comment by SumanSudhir** [comment]
After adding precondition time taken by function matmul is getting increased by around 60% when I ran it for 1M iterations as compared to matmul without precondition.

File: Sources/TensorFlow/Operators/Math.swift [hunk 2]
```diff
@@ -2748,6 +2748,9 @@ public func matmul<Scalar: Numeric>(
     _ rhs: Tensor<Scalar>,
     transposed transposeRhs: Bool = false
 ) -> Tensor<Scalar> {
+    precondition(lhs.rank >= 2, "The input must have rank greater than or equal to 2.")
+    precondition(rhs.rank >= 2, "The input must have rank greater than or equal to 2.")
+    precondition(lhs.shape[1] == rhs.shape[0] , "Dimensions must be equal,but are \(lhs.shape[1]) and \(rhs.shape[0]).")
```
Comment (Shashi456) [inline Sources/TensorFlow/Operators/Math.swift]: Nit: space after comma.

Comment (SumanSudhir) [reply Sources/TensorFlow/Operators/Math.swift]: Does space have any effect on output??

Comment (Shashi456) [reply Sources/TensorFlow/Operators/Math.swift]: Not exactly, that's just a formatting suggestion.

**Comment by SumanSudhir** [comment]
@dan-zheng can you please review and merge it so, that I can start working on adding another issue

File: Sources/TensorFlow/Operators/Math.swift (hunk 2 above)
Comment (dan-zheng) [inline Sources/TensorFlow/Operators/Math.swift]: Please keep lines under 100 columns:

```suggestion
    precondition(
      lhs.shape[1] == rhs.shape[0],
      """
      Dimension mismatch: lhs.shape[1] is not equal to rhs.shape[0] \
      (\(lhs.shape) vs \(rhs.shape)).
      """
    )
```

**Comment by dan-zheng** [comment]
> After adding precondition time taken by function matmul is getting increased by around 60% when I ran it for 1M iterations as compared to matmul without precondition.

Did you benchmark `matmul` without optimizations (`-Onone`)?

I think `-Onone` performance regressions due to adding preconditions are acceptable. Preconditions can be disabled with the `-Ounchecked` optimization mode, e.g. `swift build -c release -Xswiftc -Ounchecked`.

[This blog post](https://blog.krzyzanowskim.com/2015/03/09/swift-asserts-the-missing-manual/) has more info on which optimization modes disable assertions/preconditions.

**Comment by SumanSudhir** [comment]
I run matmul for 1M iterations with and without precondition and I found that there was an increase of 60% in run time with precondition. Here is link to [gist](https://gist.github.com/SumanSudhir/486a169b23351479d051be6d536a576c)

**Comment by dan-zheng** [comment]
It seems there's a test failure:
```
Test Case 'LayerTests.testSequential' started at 2020-01-03 23:05:26.311
Precondition failed: Dimension mismatch: lhs.shape[1] is not equal to rhs.shape[0] ([4, 1] vs [4, 1]).: file /swift-apis/Sources/TensorFlow/Operators/Math.swift, line 2753
```

That's interesting because this patch only adds preconditions without changing functional behavior. @SumanSudhir: could you please help investigate?

**Comment by SumanSudhir** [comment]
Yaa ok, I will

**Comment by SumanSudhir** [comment]
with input `([4,1] and [1,4])` even matmul without precondition is showing error


`
Fatal error: In[0] is not a matrix. Instead it has shape [2]: file /swift-base/tensorflow-swift-apis/Sources/TensorFlow/Bindings/EagerExecution.swift, line 300`

**Comment by SumanSudhir** [comment]
can you please try by removing the last dimension mismatch precondition as the matmul function takes 4 arguments with two-argument as optional. This optional argument might affect the dimension mismatch condition.

**Comment by dan-zheng** [comment]
> Can you please try by removing the last dimension mismatch precondition as the matmul function takes 4 arguments with two-argument as optional. This optional argument might affect the dimension mismatch condition.

Good point about `transposeLhs` and `transposeRhs` complicating the "dimension mismatch" precondition.

Conservatively dropping that precondition for now sounds good to me. For reference, [here's](https://github.com/tensorflow/tensorflow/blob/30b98cc73a8b046c8ce0915f6c0c7398a4947192/tensorflow/core/framework/common_shape_fns.cc#L432-L445) TensorFlow's logic for checking `matmul` inner/outer dimensions.

**Comment by SumanSudhir** [comment]
Can it be merged now?

**Comment by dan-zheng** [comment]
> Can it be merged now?

Yep!

**Comment by SumanSudhir** [comment]
Can I PR for eye function? I have written code for eye and it is similar to as in TensorFlow. I have also described it [here](https://github.com/tensorflow/swift-apis/pull/578#issuecomment-570098153)

**Comment by dan-zheng** [comment]
> Can I PR for eye function? I have written code for eye and it is similar to as in TensorFlow. I have also described it [here](https://github.com/tensorflow/swift-apis/pull/578#issuecomment-570098153)

The pending patch https://github.com/tensorflow/swift-apis/pull/578 by @Shashi456 already implements `eye` with functionality similar to your implementation. Let's try to make progress on that PR instead.

Feel free to look into [other issues](https://github.com/tensorflow/swift-apis/issues) to prevent duplicate work!

**Pull request merged**